yt/utilities/input_validator.py

```diff
@@ -0,0 +1,46 @@
+import numpy as np
+
+from yt import YTQuantity, YTArray
+from yt.data_objects.static_output import Dataset
+from yt.utilities.exceptions import YTInvalidArgumentType
+
+def input_validator(_type_name):
+    def validator(func):
+        def wrapper(*args, **kwargs):
+            bad_input = False
+            data_category_1 = (list, np.ndarray, YTArray, YTQuantity)
+            data_category_2 = (int, float, tuple, YTQuantity)
+
+            if _type_name == 'disk':
```

Comment:
I think it's unwise to divorce the validation for the signature from the signature itself. This isn't just in a separate file, it's in a separate module - which means someone could change the signature without remembering to change the validation.

I think it would be better if we could put the validation inline. Something like this in the original module (incomplete example, and my naming is probably off):

```py
from yt.utilities.input_validation import valid_array, valid_numeric

@validate_arguments(valid_array, valid_array, valid_numeric, valid_numeric)
def __init__(...):
    ...

```

In my opinion, any validation that doesn't fit cleanly like that shouldn't be in a decorator. If the decorator doesn't really work out, It's not so horrible to just put some validation into `__init__` itself, anyway. (For example, checking that argument 1 and 2 are the same length.) Having complex validation that lives in big switch statement in a different module seems like the worst of both worlds.

Reply:
I agree that putting validation in `__init__` is not so bad. However, I bet you could factor out the different types of validation we need to do into functions that we can call from each of the separate `__init__` functions. For example, you could define `validate_3d_array` which checks to see if the argument passed to is a 3D array. We already have a few functions that do things like this see specifically e.g. `validate_width_tuple` in `yt/funcs.py`. That function is helper that's used in the `__init__` of some objects in the plotting code and elsewhere in yt that processes user-supplied arguments that might be a quantity, a float, or a tuple like `(3, 'km')`.

Reply:
Arguments against the current design seems reasonable and thus I am changing as per the comments.